nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py

```diff
@@ -551,11 +548,43 @@ def _get_metrics_calculator(self) -> MetricsCalculator:
 
 
 class ActivationAPoZRankPruner(ActivationPruner):
+    def _collector(self, buffer: List) -> Callable[[Module, Tensor, Tensor], None]:
+        assert len(buffer) == 0, 'Buffer pass to activation pruner collector is not empty.'
+        # The length of the buffer used in this pruner will always be 2.
+        # buffer[0] is the number of how many batches are counted in buffer[1].
+        # buffer[1] is a tensor and the size of buffer[1] is same as the activation,
+        # each element in buffer[1] means the sum of zero counts of the same activation position cross each batch.
+        buffer.append(0)
+
+        def collect_activation(_module: Module, _input: Tensor, output: Tensor):
+            if len(buffer) == 1:
+                buffer.append(torch.zeros_like(output))
+            if buffer[0] < self.training_batches:
+                buffer[1] += torch.eq(self._activation(output.detach()), torch.zeros_like(output)).type_as(output)
+                buffer[0] += 1
+        return collect_activation
+
     def _get_metrics_calculator(self) -> MetricsCalculator:
         return APoZRankMetricsCalculator(dim=1)
 
 
 class ActivationMeanRankPruner(ActivationPruner):
+    def _collector(self, buffer: List) -> Callable[[Module, Tensor, Tensor], None]:
+        assert len(buffer) == 0, 'Buffer pass to activation pruner collector is not empty.'
+        # The length of the buffer used in this pruner will always be 2.
+        # buffer[0] is the number of how many batches are counted in buffer[1].
+        # buffer[1] is a tensor and the size of buffer[1] is same as the activation,
+        # each element in buffer[1] means the sum of the same activation position cross each batch.
+        buffer.append(0)
+
+        def collect_activation(_module: Module, _input: Tensor, output: Tensor):
+            if len(buffer) == 1:
```

Comment:
can we initialize the buffer at line-578? so that we do not need check the length of the buffer?

Reply:
I think it is hard because we need the size of the output to initialize the buffer[1].

Reply:
Get it!